nixos/modules/services/system/self-deploy.nix

```
@@ -0,0 +1,174 @@
+      type = lib.types.attrs;
+
+      default = {};
+
+      description = ''
+      Arguments to `nix-build` passed as `--argstr` or `--arg` depending on
+      the type.
+      '';
+    };
+
+    switchCommand = lib.mkOption {
+      type = lib.types.enum [ "boot" "switch" "dry-activate" "test" ];
+
+      default = "switch";
+
+      description = ''
+      The `switch-to-configuration` subcommand used.
+      '';
+    };
+
+    repository = lib.mkOption {
+      type = with lib.types; oneOf [ path str ];
+
+      description = ''
+      The repository to fetch from. Must be properly formatted for git.
+
+      If this value is set to a path (must begin with `/`) then it's
+      assumed that the repository is local and the resulting service
+      won't wait for the network to be up.
+
+      If the repository will be fetched over SSH, you must add an
+      entry to `programs.ssh.knownHosts` for the SSH host for the fetch
+      to be successful.
+      '';
+    };
+
+    sshKeyFile = lib.mkOption {
+      type = with lib.types; nullOr path;
+
+      default = null;
+
+      description = ''
+      Path to SSH private key used to fetch private repositories over
+      SSH.
+      '';
+    };
+
+    branch = lib.mkOption {
+      type = lib.types.str;
+
+      default = "master";
+
+      description = ''
+      Branch to track
+
+      You can also specify a revision or tag, too, if you want to pin this
+      machine to a specific commit.
+      '';
+    };
+
+    startAt = lib.mkOption {
+      type = with lib.types; either str (listOf str);
+
+      default = "hourly";
+
+      description = ''
+      The schedule on which to run the `self-deploy` service. Format
+      specified by `systemd.time 7`.
+
+      This value can also be a list of `systemd.time 7` formatted
+      strings, in which case the service will be started on multiple
+      schedules.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.self-deploy = {
+      wantedBy = [ "multi-user.target" ];
+
+      requires = lib.mkIf (!(isPathType cfg.repository)) [ "network-online.target" ];
+
+      environment.GIT_SSH_COMMAND = lib.mkIf (!(isNull cfg.sshKeyFile))
+        "${pkgs.openssh}/bin/ssh -i ${lib.escapeShellArg cfg.sshKeyFile}";
+
+      serviceConfig.X-RestartIfChanged = false;
+
+      path = with pkgs; [
+        git nix systemd
+      ];
+
+      script = ''
+      if [ ! -e ${workingDirectory} ]; then
+        mkdir --parents ${workingDirectory}
+      fi
+
+      if [ ! -e ${repositoryDirectory} ]; then
+        git clone ${lib.cli.toGNUCommandLineShell {} {
+          local = (isPathType cfg.repository);
+        }} ${lib.escapeShellArg cfg.repository} ${repositoryDirectory}
+      fi
+
+      # Ensure that if cfg.repository is changed, origin is updated
+      ${gitWithRepo} remote set-url origin ${lib.escapeShellArg cfg.repository}
+
+      ${gitWithRepo} fetch origin ${lib.escapeShellArg cfg.branch}
```

I think you can simplify this and make it more "pure" by doing this:

```suggestion
      ${gitWithRepo} fetch ${lib.escapeShellArg cfg.repository} ${lib.escapeShellArg cfg.branch}
```

---

Hmm, so it doesn't need to be added as a remote to fetch from it? interesting

I wonder if it would be better to create a bare git repository instead of cloning in that case

---

Confirmed the fresh git repo works, going to switch to that